### DISCUSSION TITLE
How does Chezmoi respond to symlinks inside the Chezmoi root?

### DISCUSSION THREAD

**Discussion by kartikynwa** [twpayne/chezmoi#3349] [Q&A] [answered]
I started wondering this after trying to add Windows support to my pre-existing Linux Chezmoi repository. Mainly, I had an emacs configuration folder in `~/.config/emacs` for Linux. I wanted to have the exact same configuration in Windows except at a different path `%APPDATA%\Roaming\.emacs.d`.

I am aware that there is documentation for handling cases like this [HERE](https://www.chezmoi.io/user-guide/manage-machine-to-machine-differences/#handle-different-file-locations-on-different-systems-with-the-same-contents). But I naively first tried to create a symlink like so along with using `.chezmoiignore` to conditionally copy only the files relevant to the OS (here `.chezmoiroot` is `dotfiles`):

```
mkdir -p dotfiles/AppData/Roaming
cd dotfiles/AppData/Roaming
ln -s ../../dot_config/emacs dot_emacs.d
```

This only causes the creation of an empty folder at `%APPDATA%\Roaming\.emacs.d`.

So I started wondering. How does Chezmoi understand symlinks inside the `chezmoiroot`?

**Comment by twpayne** [comment, marked as answer]
chezmoi follows symlinks to **files** in the source directory but **not** symlinks to **directories**.

The reason that chezmoi does not follow links to directories is that doing so would make it easy to get stuck in an infinite loop (imagine if a symbolic link in a subdirectory pointed to one of the subdirectory's parents).

The short-term work around for you is to replace the link to the `dot_emacs.d` directory with a link for each file inside the directory instead.

In the long term using symbolic links to directories is alternative to #2273 is an interesting idea, but will require chezmoi to detect and respond to links to directories itself to avoid getting stuck in the aforementioned infinite loop. This would also break chezmoi's promise that it only requires regular files and directories in your source directory (Windows, by default, doesn't support symbolic links, nor do some version control systems). So, #2273 will probably, eventually, get implemented instead.

BTW I see you have a [two-stage `chezmoi apply` in your dotfiles](https://github.com/kartikynwa/dotfiles/blob/33a6f9309bb3061f1d156290420dba1b3a1f7d52/README.md#apply-the-dotfiles). You should use a [config file template](https://www.chezmoi.io/user-guide/setup/#re-create-your-config-file) instead.

**Reply by kartikynwa** [reply]
> BTW I see you have a [two-stage chezmoi apply in your dotfiles](https://github.com/kartikynwa/dotfiles/blob/33a6f9309bb3061f1d156290420dba1b3a1f7d52/README.md#apply-the-dotfiles). You should use a [config file template](https://www.chezmoi.io/user-guide/setup/#re-create-your-config-file) instead.

Thanks for this.
